File: leetcode-cpp/solutions/0001_two_sum.cpp
#include <vector>
#include <unordered_map>
using std::vector;

class Solution {
public:
    /**
     * 两数之和 - 哈希表解法
     * 时间复杂度：O(n)
     * 空间复杂度：O(n)
     */
    vector<int> twoSum(vector<int>& nums, int target) {
        // 哈希表：key存储数组元素值，value存储该元素的索引
        std::unordered_map<int, int> myMap;

        // 遍历数组
        for (int i = 0; i < nums.size(); i++)
        {
            // 计算目标差值，并在哈希表中查找
            auto iter = myMap.find(target - nums[i]);

            if (iter != myMap.end())  // 找到配对的数
            {
                // 返回两个索引：之前找到的数的索引 和 当前索引
                return { iter->second, i };
            }
            else  // 未找到配对
            {
                // 将当前元素及其索引存入哈希表，供后续查找使用
                myMap.insert(std::pair<int, int>(nums[i], i));
            }
        }

        return {};
    }
};

File: leetcode-python/solutions/0001_two_sum.py
class Solution:
    """
    两数之和 - 哈希表解法
    时间复杂度: O(n)
    空间复杂度: O(n)
    """
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        # 创建哈希表，用于存储 {数值: 索引} 的映射关系
        records = dict()

        # 遍历数组，index 为当前索引，value 为当前值
        for index, value in enumerate(nums):
            # 计算差值：target - value
            if target - value in records:
                # 返回两个数的索引：[之前记录的索引, 当前索引]
                return [records[target - value], index]

            # 将当前数值及其索引存入哈希表，供后续查找使用
            records[value] = index

        return []